### ISSUE TITLE
(t) Refresh or remove as required all test fixtures

### ISSUE THREAD

**Issue by phillxnet** [rockstor/rockstor-core#2384]
(t) = testing branch. Prior to the work begun in '(t) Failing unit tests' #2371 as a result of major Django changes enacted in 'Update Django within current Python 2 constrains #2254' #2365, and following on from the fixture re-enablement spin-off of the former: '(t) Test fixtures non functional #2382' #2383 we have a test wide requirement to update or remove our now mostly outdated and newely re-enabled fixtures to address our current test situation of:

```
Ran 86 tests in 4.514s

FAILED (failures=3, errors=29)
```

in comparison to our current stable channel result of:
```
Ran 217 tests in 21.718s

OK
```

It is proposed that this issue be at minimum an address of the fixtures removal/update. Some test editing and mock object removal will inevitably be required especially where we don't just remove the fixtures that were not 'in-play' anyway in our last stable, and end up replacing the mocks and hand-created objects with the more robust fixture arrangement anyway. #2371 is evidence of a requirement to do this in some areas.

## Design goals
1. Fixtures to have an absolute minimum content.
2. Fixture reuse acceptable, thought discouraged as it can often challenging 1.
3. Always provide exact and proven command line instructions to reproduce any fixture used.
4. Always provide exact and proven command line instructions to run each test in isolation.

Where 3. & 4. are within the respective test files as comments towards the top of the file.

**Comment by phillxnet** [comment]
@FroggyFlox Wow thanks; that's a nice break down. I'll get chipping away.

**Comment by FroggyFlox** [comment]
> `AssertionError`: raw HTML returned as error message, test update likely.

These are all occurring in `test_task_scheduler.py` and very likely relate to #2370.
Removing the trailing slash in the URL of the `post` tests in this class fixes the errors:
```
buildvm:/opt/rockstor-dev # bin/test -p "test_task_scheduler.py"
Creating test database for alias 'default'...
Creating test database for alias 'smart_manager'...
System check identified no issues (0 silenced).
........
----------------------------------------------------------------------
Ran 8 tests in 0.983s

OK
Destroying test database for alias 'default'...
Destroying test database for alias 'smart_manager'...
```

**Comment by phillxnet** [comment]
Re:

> These are all occurring in `test_task_scheduler.py` and very likely relate to #2370.
> Removing the trailing slash in the URL of the `post` tests in this class fixes the errors:

Pre:
```
test_post_invalid_type (smart_manager.tests.test_task_scheduler.TaskSchedulerTests) ... FAIL
test_post_name_exists (smart_manager.tests.test_task_scheduler.TaskSchedulerTests) ... FAIL
test_post_valid (smart_manager.tests.test_task_scheduler.TaskSchedulerTests) ... FAIL
...
----------------------------------------------------------------------
Ran 8 tests in 0.658s

FAILED (failures=3)
```
Post pending commit with your suggestions:

```
test_delete_invalid (smart_manager.tests.test_task_scheduler.TaskSchedulerTests) ... ok
test_delete_valid (smart_manager.tests.test_task_scheduler.TaskSchedulerTests) ... ok
test_get (smart_manager.tests.test_task_scheduler.TaskSchedulerTests) ... ok
test_post_invalid_type (smart_manager.tests.test_task_scheduler.TaskSchedulerTests) ... ok
test_post_name_exists (smart_manager.tests.test_task_scheduler.TaskSchedulerTests) ... ok
test_post_valid (smart_manager.tests.test_task_scheduler.TaskSchedulerTests) ... ok
test_put_invalid (smart_manager.tests.test_task_scheduler.TaskSchedulerTests) ... ok
test_put_valid (smart_manager.tests.test_task_scheduler.TaskSchedulerTests) ... ok

----------------------------------------------------------------------
Ran 8 tests in 0.743s

OK
```
Oh year. Now confirmed and committed in issue branch with attribution.

**Comment by phillxnet** [comment]
Proposed fix for IntegrityError example test_disks.py:
```
test_blink_drive (storageadmin.tests.test_disks.DiskTests) ... ok
test_btrfs_disk_import_fail (storageadmin.tests.test_disks.DiskTests) ... ok
test_disable_smart (storageadmin.tests.test_disks.DiskTests) ... ok
test_disk_scan (storageadmin.tests.test_disks.DiskTests) ... ok
test_disk_wipe (storageadmin.tests.test_disks.DiskTests) ... ok
test_enable_smart (storageadmin.tests.test_disks.DiskTests) ... ok
test_enable_smart_when_available (storageadmin.tests.test_disks.DiskTests) ... ok
test_invalid_command (storageadmin.tests.test_disks.DiskTests) ... ok
test_invalid_disk_wipe (storageadmin.tests.test_disks.DiskTests) ... ok

----------------------------------------------------------------------
Ran 9 tests in 0.256s

OK
```
details in associated commit. We essentially return to what we had, i.e. no fixture, bar what is now needed for auth.

**Comment by phillxnet** [comment]
Pending commit results for the remaining IntegrityError examples via outdated and now removed fixtures and legacy inheritance:
```
Ran 8 tests in 0.658s

FAILED (failures=3)
```
Post pending commit with your suggestions:

```
bin/test -v 2 -p "test_email_client.py"

test_delete_requests (storageadmin.tests.test_email_client.EmailTests) ... ok
test_get (storageadmin.tests.test_email_client.EmailTests) ... ok
test_post_requests_1 (storageadmin.tests.test_email_client.EmailTests) ... ok
test_post_requests_2 (storageadmin.tests.test_email_client.EmailTests) ... ok

----------------------------------------------------------------------
Ran 4 tests in 0.069s

OK
```

**Comment by phillxnet** [comment]
Pending commit result:
Next wave is top level (from original summary) of DeserializationError, again we are to just remove the fixture adding our new independant auth only fixture in it's place.
```
bin/test -v 2 -p "test_commands.py"
Ran 17 tests in 2.183s
OK
```

**Comment by phillxnet** [comment]
Current summary of status via prior grep method with all published commits in place:
```
bin/test > test-results.txt 2>&1
grep "Error:" test-results.txt | cut -d ":" -f 1 | sort | uniq -c | sort -rnk 1
      6 IntegrityError
      2 AssertionError
      1 InterfaceError
      1 CommandError
```
However we now have:
```
Ran 206 tests in 18.588s

FAILED (failures=2, errors=8)
```

**Comment by phillxnet** [comment]
```
bin/test > test-results.txt 2>&1
grep "Error:" test-results.txt | cut -d ":" -f 1 | sort | uniq -c | sort -rnk 1
      6 IntegrityError
      2 AssertionError
      1 InterfaceError
```

**Comment by FroggyFlox** [comment]
> Note that when I tested `test_snmp.py` earlier today, I got different error depending on whether it was run by itself (`bin/test -p "test_snmp.py"`) or along all other tests (`bin/test`)...

@phillxnet , I was wrong about that one... I am actually seeing this oddity with `test_task_scheduler.py`.
After a bit of looking, this seems to be due to the test ran previously that is not closing properly... In this case, the tests in `test_task_scheduler.py` are failing because we are not properly setting up the two classes of tests related to `smart_manager`. These are:
- `test_snmp.py`
- `test_task_scheduler.py`

Whether we need `APITestMixin` here is still unsure, but the `connection is already closed` error is probably related to:
https://docs.djangoproject.com/en/1.11/topics/testing/tools/#django.test.SimpleTestCase.allow_database_queries
> SimpleTestCase and its subclasses (e.g. TestCase, …) rely on setUpClass() and tearDownClass() to perform some class-wide initialization (e.g. overriding settings). If you need to override those methods, don’t forget to call the super implementation:

Inheriting from `APITestMixin` and setting up the `setUpClass()` & `tearDownClass()` with proper `super()` calls within each seems to fix the issue.

**Comment by phillxnet** [comment]
@FroggyFlox Re:

> Inheriting from `APITestMixin` and setting up the `setUpClass()` & `tearDownClass()` with proper `super()` calls within each in `test_snmp.py` and `test_task_scheduler.py` seems to fix the issue.

Nice find. I'll pop this in and take a look, wont be before tomorrow now. That "connection is already closed" was quite a strange one so thanks for looking into that. We should be able to flush out all the existing strangeness during this issue and we are now down to a manageable number of errors/failures. With most now likely fixed by moving to proper fixtures.

**Comment by FroggyFlox** [comment]
> .. I think we can still load the fixtures (and thus address #2382),

Confirmed by adding the following small test to `test_pool_scrub.py`:
```python
    def test_read_fixture(self):
        p = Pool.objects.get(pk=2)
        self.assertEqual(p.name, 'rock-pool')
```
and it still works:
```
test_read_fixture (storageadmin.tests.test_pool_scrub.PoolScrubTests) ... ok
```

**Comment by phillxnet** [comment]
Quick note on last test_appliance.py commit here: replace the db object mocking in test_appliances.py with now working fixtures (thanks again @FroggyFlox ) we now have a full-house on the also simplified test_appliances.py test set:
```
Ran 4 tests in 0.071s
OK

Where as before our test_get was not responding as it had done previous to our recent Django update.
```

@FroggyFlox Thanks for the updated pointers on a better fix for #2382, as discussed in side channel I'll pop them in here as I go.

Results so far with commits to date:
```
Ran 209 tests in 18.484s
FAILED (failures=1, errors=7)
with test_task_scheduler.py having its in-context with other tests:
> InterfaceError: Problem installing fixture '/opt/rockstor-dev/src/rockstor/smart_manager/fixtures/scheduled_tasks.json': connection already closed
But running find on it's own.
```
and post above @FroggyFlox suggested re-work of the recent #2382 fix.
```
Ran 217 tests in 18.988s
FAILED (failures=1, errors=13)
```
Where we are now down to simple fixture related issues by the looks of it.
Also noteworthy here is that our suspect test_appliances.py test is now running just dandy in the contest of the other tests.
Likely confirming @FroggyFlox fix here. @FroggyFlox if you could possible take a look at this last commit to ensure I have what you had intended within it that, when you get the time. Cheers.

**Comment by FroggyFlox** [comment]
> @FroggyFlox if you could possible take a look at this last commit to ensure I have what you had intended within it that, when you get the time. Cheers.

Looks exactly like what I've tested, thanks a lot for including it!

**Comment by phillxnet** [comment]
Having now moved from a pure mock db setup to a pure fixtures setup in test_disk_smart.py
we go from:
```
./bin/test -v 2 -p test_disk_smart.py

test_get (storageadmin.tests.test_disk_smart.DiskSmartTests) ... ok
test_post_reqeusts_1 (storageadmin.tests.test_disk_smart.DiskSmartTests) ... ok
test_post_requests_2 (storageadmin.tests.test_disk_smart.DiskSmartTests) ... FAIL

======================================================================
FAIL: test_post_requests_2 (storageadmin.tests.test_disk_smart.DiskSmartTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/opt/rockstor-dev/src/rockstor/storageadmin/tests/test_disk_smart.py", line 133, in test_post_requests_2
    self.assertEqual(response.status_code, status.HTTP_200_OK, msg=response.data)
AssertionError: ['Failed to insert expression "<MagicMock name=\'info().__getitem__().resolve_expression()\' id=\'139929588376720\'>" on storageadmin.SMARTIdentity.model_family. F() expressions can only be used to update, not to insert.'
...
ValueError: Failed to insert expression "<MagicMock name=\'info().__getitem__().resolve_expression()\' id=\'139929588376720\'>" on storageadmin.SMARTIdentity.model_family. F() expressions can only be used to update, not to insert.\n' ...

Ran 3 tests in 0.148s

FAILED (failures=1)
```
to:
```
test_get (storageadmin.tests.test_disk_smart.DiskSmartTests) ... ok
test_post_reqeusts_1 (storageadmin.tests.test_disk_smart.DiskSmartTests) ... ok
test_post_requests_2 (storageadmin.tests.test_disk_smart.DiskSmartTests) ... ok

----------------------------------------------------------------------
Ran 3 tests in 0.647s

OK
```
No other functional change.

**Comment by phillxnet** [comment]
Status thus far:
```
...
test_delete (storageadmin.tests.test_network.NetworkTests) ... ERROR
test_get_base (storageadmin.tests.test_network.NetworkTests) ... ERROR
test_nclistview_post_devices (storageadmin.tests.test_network.NetworkTests) ... ERROR
test_nclistview_post_devices_not_list (storageadmin.tests.test_network.NetworkTests) ... ERROR
test_nclistview_post_invalid (storageadmin.tests.test_network.NetworkTests) ... ERROR
test_put (storageadmin.tests.test_network.NetworkTests) ... ERROR
test_put_invalid_id (storageadmin.tests.test_network.NetworkTests) ... ERROR
...
ERROR: test_create_samba_share (storageadmin.tests.test_samba.SambaTests)
ERROR: test_post_requests_2 (storageadmin.tests.test_samba.SambaTests)
ERROR: test_post_requests_no_admin (storageadmin.tests.test_samba.SambaTests)
ERROR: test_put_requests_2 (storageadmin.tests.test_samba.SambaTests)
ERROR: test_clone_command (storageadmin.tests.test_snapshot.SnapshotTests)
ERROR: test_post_requests_2 (storageadmin.tests.test_snapshot.SnapshotTests)

Ran 217 tests in 20.756s

FAILED (errors=13)
OK
```

**Comment by phillxnet** [comment]
Pre fix on test_snapshot.py
```
test_clone_command (storageadmin.tests.test_snapshot.SnapshotTests) ... ok
ERROR - (IntegrityError) DETAIL:  Key (pool_id)=(1) is not present in table "storageadmin_pool".
test_delete_requests (storageadmin.tests.test_snapshot.SnapshotTests) ... ok
test_get (storageadmin.tests.test_snapshot.SnapshotTests) ... ok
test_post_requests_1 (storageadmin.tests.test_snapshot.SnapshotTests) ... ok
test_post_requests_2 (storageadmin.tests.test_snapshot.SnapshotTests) ... ok
ERROR -  (IntegrityError) DETAIL:  Key (pool_id)=(1) is not present in table "storageadmin_pool".


Ran 5 tests in 0.198s
FAILED (errors=2)
```

**Comment by FroggyFlox** [comment]
I was in the same state of mind and that's why I tried to have a go at it. It turned out to be not too bad.
The difficult part will be to get it back to its "prior" prior state as we had to comment a lot of tests that were failing but that deserves its own issue.

**Comment by phillxnet** [comment]
Quick issue branch status update, including the cherry picked commit not yet visible publicly in issue branch but referenced by @FroggyFlox 
```
bin/test -v 2

Ran 217 tests in 21.289s

FAILED (errors=2)

OK
```
With the 2 remaining errors relating to my current mid-task concerning the move to fixtures for test_snapshot.py referenced in comment: https://github.com/rockstor/rockstor-core/issues/2384#issuecomment-1164749369 above.

**Comment by FroggyFlox** [comment]
@phillxnet, I noticed one discrepancy with your results on `test_disk_smart.py`. I, indeed, receive the following:
```
test_get (storageadmin.tests.test_disk_smart.DiskSmartTests) ... ok
test_post_reqeusts_1 (storageadmin.tests.test_disk_smart.DiskSmartTests) ... ok
test_post_requests_2 (storageadmin.tests.test_disk_smart.DiskSmartTests) ... FAIL

======================================================================
FAIL: test_post_requests_2 (storageadmin.tests.test_disk_smart.DiskSmartTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/opt/rockstor-dev/eggs/mock-1.0.1-py2.7.egg/mock.py", line 1201, in patched
    return func(*args, **keywargs)
  File "/opt/rockstor-dev/src/rockstor/storageadmin/tests/test_disk_smart.py", line 117, in test_post_requests_2
    self.assertEqual(response.status_code, status.HTTP_200_OK, msg=response.data)
AssertionError: ["Error running a command. cmd = /usr/sbin/smartctl -t short /dev/disk/by-id/ata-QEMU_HARDDISK_QM00003. rc = 1. stdout = ['smartctl 7.2 2021-09-14 r5237 [x86_64-linux-5.3.18-150300.59.71-default] (SUSE RPM)', 'Copyright (C) 2002-20, Bruce Allen, Christian Franke, www.smartmontools.org', '', '/dev/disk/by-id/ata-QEMU_HARDDISK_QM00003: Unable to detect device type', 'Please specify device type with the -d option.', '', 'Use smartctl -h to get a usage summary', '', '']. stderr = ['']", 'Traceback (most recent call last):\n  File "/opt/rockstor-dev/src/rockstor/rest_framework_custom/generic_view.py", line 41, in _handle_exception\n    yield\n  File "/opt/rockstor-dev/src/rockstor/storageadmin/views/disk_smart.py", line 176, in post\n    run_test(disk.name, test_type, disk.smart_options)\n  File "/opt/rockstor-dev/src/rockstor/system/smart.py", line 325, in run_test\n    o, e, rc = run_command([SMART, "-t", test] + get_dev_options(device, custom_options))\n  File "/opt/rockstor-dev/src/rockstor/system/osi.py", line 225, in run_command\n    raise CommandException(cmd, out, err, rc)\nCommandException: Error running a command. cmd = /usr/sbin/smartctl -t short /dev/disk/by-id/ata-QEMU_HARDDISK_QM00003. rc = 1. stdout = [\'smartctl 7.2 2021-09-14 r5237 [x86_64-linux-5.3.18-150300.59.71-default] (SUSE RPM)\', \'Copyright (C) 2002-20, Bruce Allen, Christian Franke, www.smartmontools.org\', \'\', \'/dev/disk/by-id/ata-QEMU_HARDDISK_QM00003: Unable to detect device type\', \'Please specify device type with the -d option.\', \'\', \'Use smartctl -h to get a usage summary\', \'\', \'\']. stderr = [\'\']\n']
```

It seems the mock for `run_test()` is not working and the smart test tries to actually run.
I could have a seemingly functional mock when changing the `run_test()` mock to:
```python
    @classmethod
    def setUpClass(cls):
        super(DiskSmartTests, cls).setUpClass()

        # post mock
        cls.patch_run_test = patch("system.smart.run_test")
        cls.mock_run_test = cls.patch_run_test.start()
        cls.mock_run_test.return_value = [""], [""], 0
```
results:
```
test_get (storageadmin.tests.test_disk_smart.DiskSmartTests) ... ok
test_post_reqeusts_1 (storageadmin.tests.test_disk_smart.DiskSmartTests) ... ok
test_post_requests_2 (storageadmin.tests.test_disk_smart.DiskSmartTests) ... ok

----------------------------------------------------------------------
Ran 3 tests in 0.075s

OK
```

**Comment by phillxnet** [comment]
@FroggyFlox Cheers, Nice find. I'm still obviously testing on the system I'm creating fixtures etc on so good to have this failing mock found. I've committed appropriately to get at least this one out of the way.

**Comment by phillxnet** [comment]
Remarking out the entirety of test_shares.py, the test run directly before our 'sensitive' test_snapshot.py results in it executing in context just fine:
```
test_get (storageadmin.tests.test_sftp.SFTPTests) ... ok
test_post_requests_1 (storageadmin.tests.test_sftp.SFTPTests) ... ok
test_post_requests_2 (storageadmin.tests.test_sftp.SFTPTests) ... ok
test_clone_command (storageadmin.tests.test_share_commands.ShareCommandTests) ... ok
test_rollback_command (storageadmin.tests.test_share_commands.ShareCommandTests) ... ok
test_clone_command (storageadmin.tests.test_snapshot.SnapshotTests) ... ok
test_delete_requests (storageadmin.tests.test_snapshot.SnapshotTests) ... ok
test_get (storageadmin.tests.test_snapshot.SnapshotTests) ... ok
test_post_requests_1 (storageadmin.tests.test_snapshot.SnapshotTests) ... ok
test_post_requests_2 (storageadmin.tests.test_snapshot.SnapshotTests) ... ok
test_get (storageadmin.tests.test_tls_certificate.TlscertificateTests) ... ok
test_post_requests (storageadmin.tests.test_tls_certificate.TlscertificateTests) ... ok
test_get (storageadmin.tests.test_update_subscription.UpdateSubscriptionTests) ... ok
```
And although this does not approach the issue of cross test object mock contamination it does narrow down the test that is 'leaking'. I'll take a look now at moving test_shares.py over to the fixture mock method and remove as many, if not all, of the db object mocking done in that test as they may well be reaching out to mocking objects in all tests so would still represent a fix.

**Comment by phillxnet** [comment]
Post moving test_share.py to newly functional fixture based db instantiation we have some more descriptive test group names and one additional test group:
```
./bin/test -v 2 -p test_shares.py

test_compression (storageadmin.tests.test_shares.ShareTests) ... ok
test_create (storageadmin.tests.test_shares.ShareTests) ... ok
test_delete_exported_replicated (storageadmin.tests.test_shares.ShareTests) ... ok
test_delete_no_exports_services_snaps (storageadmin.tests.test_shares.ShareTests) ... ok
test_delete_os_exception (storageadmin.tests.test_shares.ShareTests) ... ok
test_delete_rock_ons_root (storageadmin.tests.test_shares.ShareTests) ... ok
test_delete_with_regular_snapshot (storageadmin.tests.test_shares.ShareTests) ... ok
test_get (storageadmin.tests.test_shares.ShareTests) ... ok
test_name_regex (storageadmin.tests.test_shares.ShareTests) ... ok
test_resize (storageadmin.tests.test_shares.ShareTests) ... ok

----------------------------------------------------------------------
Ran 10 tests in 3.014s

OK
```
And we have addressed the prior mocking leaky of test_shares.py into test_snapshot.py
```
...
test_name_regex (storageadmin.tests.test_shares.ShareTests) ... ok
test_resize (storageadmin.tests.test_shares.ShareTests) ... ok
test_clone_command (storageadmin.tests.test_snapshot.SnapshotTests) ... ok
test_delete_requests (storageadmin.tests.test_snapshot.SnapshotTests) ... ok
test_get (storageadmin.tests.test_snapshot.SnapshotTests) ... ok
test_post_requests_1 (storageadmin.tests.test_snapshot.SnapshotTests) ... ok
test_post_requests_2 (storageadmin.tests.test_snapshot.SnapshotTests) ... ok
test_get (storageadmin.tests.test_tls_certificate.TlscertificateTests) ... ok
...


Ran 218 tests in 22.139s

OK
 OK
```

**Comment by phillxnet** [comment]
I've now deleted and re-based my working branch associated with this issue onto current test post (#2387) and squashed all commits to date, preserving cherry picked commits from other contributors (@FroggyFlox) thus far.
Still a work in progress but getting there now.

**Comment by phillxnet** [comment]
Just now removed GitHub resident instance of working branch, re-based after merging #2389, and re-pushed post squashing. Hopefully this will be the last time this working branch has to be re-based due to spin-off issues/pull requests.

**Comment by phillxnet** [comment]
Current status of working branch associated with this issue:
```
...
FAIL: test_post_requests_2 (storageadmin.tests.test_disk_smart.DiskSmartTests) - we have an outstanding mocking issue
...
FAIL: test_raid1_crud (storageadmin.tests.test_pools.PoolTests) - ongoing BoolBalance db manipulation issues 
...
Ran 224 tests in 31.974s

FAILED (failures=2)
 OK
```
So with the ongoing re-arrangement we now have a few more tests and those to-date are now all the more relevant.
Getting there.

**Comment by phillxnet** [comment]
Status thus far:
```
FAIL: test_post_requests_2 (storageadmin.tests.test_disk_smart.DiskSmartTests) - we have an outstanding mocking issue
...
Ran 225 tests in 31.705s

FAILED (failures=1)
 OK
```
About to establish some new test to look further into a suspected PoolBalance status reporting bug. If successful these tests can then be the reproducer for a spin-off issue regarding false positives for ongoing pool balances.

**Comment by phillxnet** [comment]
Re-based working branch post #2391

**Comment by phillxnet** [comment]
Closing as:
Fixed by #2400
